docs/r3/C5-Ethereum-Interop-Architecture.md

```diff
@@ -0,0 +1,379 @@
+
+  * [https://github\.com/ethereum/execution\-apis/tree/main](https://github.com/ethereum/execution-apis/tree/main)
+
+# Outbound Communication
+
+## How to use Web3?
+
+What mechanisms are in place to allow us to use Web3 in Corda?
+
+* __Web3j__  – is an industry standard library approach that gives us:
+  * The ability to transform Solidity contracts to Java
+  * A mechanism for transmitting that information to an EVM network
+  * Well maintained \(and Apache 2\.0\)
+  * Already used by R3 \(in demos\)
+
+![](img/ExtInterop_C5ARB2.png)
+![](img/ExtInterop_C5ARB1.png)
+
+## Web3j in a flow?!
+
+This makes the code, from within the flow the same as what would be used by any java developer.
+
+
+```
+private val web3j: Web3j
+
+val contract = ERC20\.load\(address\, web3j\, …\)
+
+val balance = contract\.balanceOf\(…\)
+```
+
+But…
+
+It’s not `OSGi` compliant (so we can’t use in a sandbox).  Also it tends to prefer direct access to a chain (e.g., via HTTP).  And then there are transitive dependencies…
+
+## Solution(a) Make Web3j work for us
+
+  * We _can_ make it work with `OSGi`
+  * We _can_ get around the HTTP connection by injecting a proxy for communication (Web3j is designed specifically to allow this)
+  * We can manage the dependencies
+
+The problem is that all of this could be difficult to maintain.
+
+## Solution(b) Encode the Web3 functions ourselves
+
+A compiled smart contract will provide a `json` representation of the smart contract functions called the ABI ([Application Binary Interface](https://docs.soliditylang.org/en/latest/abi-spec.html#)), e.g.:
+
+```
+"abi": [
+  {
+    "inputs": [
+      {
+        "internalType": "address",
+        "name": "newOwner",
+        "type": "address"
+      }
+    ],
+    "name": "changeOwner",
+    "outputs": [],
+    "stateMutability": "nonpayable",
+    "type": "function"
+  },
+  {
+    "inputs": [],
+    "name": "getOwner",
+    "outputs": [
+      {
+        "internalType": "address",
+        "name": "",
+        "type": "address"
+      }
+    ],
+    "stateMutability": "view",
+    "type": "function"
+  }
+]
+```
+
+* We can do the same as Web3j and encode the Web3 function ourselves.
+  * Structure for functions can be read from compiled solidity code
+  * Removes the need for `OSGi`
+  * No new dependencies
+
+## Recap - A basic primer on the architecture of Corda 5
+
+Before we continue it's worth revisiting the architecture of Corda 5.  Corda 5 operates as a cluster or [workers](https://docs.r3.com/en/platform/corda/5.0/key-concepts/cluster-admin/workers.html), each of which may contain one or more `processors`.  
+
+The workers are the runtime applications, while the processors are what do the work.  It is possible, and even likely, that we will have multiple workers running at a time, allowing for some parallelism of the processors, giving us the ability to execute multiple instructions (in our case, EVM commands) at the same time.

```

Comment:
is there a danger that 'application' is ambiguous here? (ie these are not the 'applications' that a customer writes, right?  although they may execute some customer code). Perhaps we call them the 'runtime services' or 'runtime components' of Corda?

Reply:
I'll use services.  Thanks